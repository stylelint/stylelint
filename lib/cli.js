#!/usr/bin/env node
/* @flow */
"use strict";
const dynamicRequire = require("./dynamicRequire");
const getModulePath = require("./utils/getModulePath");
const getStdin = require("get-stdin");
const meow = require("meow");
const needlessDisablesStringFormatter = require("./formatters/needlessDisablesStringFormatter");
const path = require("path");
const resolveFrom = require("resolve-from");
const standalone = require("./standalone");

const minimistOptions = {
  default: {
    config: false,
    f: "string",
    q: false
  },
  alias: {
    f: "formatter",
    h: "help",
    i: "ignore-path",
    ip: "ignore-pattern",
    id: "ignore-disables",
    q: "quiet",
    rd: "report-needless-disables",
    s: "syntax",
    v: "version",
    di: "disable-default-ignores"
  },
  boolean: [
    "color",
    "help",
    "ignore-disables",
    "no-color",
    "quiet",
    "version",
    "disable-default-ignores",
    "fix"
  ]
};

const meowOptions = {
  help: `
    Usage: stylelint [input] [options]

    Input: Files(s), glob(s), or nothing to use stdin.

      If an input argument is wrapped in quotation marks, it will be passed to
      node-glob for cross-platform glob support. node_modules and
      bower_components are always ignored. You can also pass no input and use
      stdin, instead.

    Options:

      --config

        Path to a specific configuration file (JSON, YAML, or CommonJS), or the
        name of a module in node_modules that points to one. If no --config
        argument is provided, stylelint will search for configuration files in
        the following places, in this order:
          - a stylelint property in package.json
          - a .stylelintrc file (with or without filename extension:
            .json, .yaml, .yml, and .js are available)
          - a stylelint.config.js file exporting a JS object
        The search will begin in the working directory and move up the directory
        tree until a configuration file is found.

      --config-basedir

        An absolute path to the directory that relative paths defining "extends"
        and "plugins" are *relative to*. Only necessary if these values are
        relative paths.

      --ignore-path, -i

        Path to a file containing patterns that describe files to ignore. The
        path can be absolute or relative to process.cwd(). By default, stylelint
        looks for .stylelintignore in process.cwd().

      --ignore-pattern, -ip

        Pattern of files to ignore (in addition to those in .stylelintignore)

      --syntax, -s

        Specify a non-standard syntax. Options: "scss", "less", "sugarss".
        If you do not specify a syntax, non-standard syntaxes will be
        automatically inferred by the file extensions .scss, .less, and .sss.

      --fix

        Automatically fix violations of certain rules.

      --custom-syntax

        Module name or path to a JS file exporting a PostCSS-compatible syntax.

      --stdin-filename

        A filename to assign stdin input.

      --ignore-disables, --id

        Ignore styleline-disable comments.

      --disable-default-ignores, --di

        Allow linting of node_modules and bower_components.

      --cache                       [default: false]

        Store the info about processed files in order to only operate on the
        changed ones the next time you run stylelint. By default, the cache
        is stored in "./.stylelintcache". To adjust this, use --cache-location.

      --cache-location              [default: '.stylelintcache']

        Path to a file or directory to be used for the cache location.
        Default is "./.stylelintcache". If a directory is specified, a cache
        file will be created inside the specified folder, with a name derived
        from a hash of the current working directory.

        If the directory for the cache does not exist, make sure you add a trailing "/"
        on *nix systems or "\\" on Windows. Otherwise the path will be assumed to be a file.

      --formatter, -f               [default: "string"]

        The output formatter: "json", "string" or "verbose".

      --custom-formatter

        Path to a JS file exporting a custom formatting function.

      --quiet, -q

        Only register warnings for rules with an "error"-level severity (ignore
        "warning"-level).

      --color
      --no-color

        Force enabling/disabling of color.

      --allow-empty-input, --aei

        If no files match glob pattern, exits without throwing an error.

      --report-needless-disables, --rd

        Report stylelint-disable comments that are not blocking a lint warning.
        If you provide the argument "error", the process will exit with code 2
        if needless disables are found.

      --version, -v

        Show the currently installed version of stylelint.
  `,
  pkg: "../package.json"
};

const cli = meow(meowOptions, minimistOptions);

let formatter = cli.flags.formatter;
if (cli.flags.customFormatter) {
  const customFormatter = path.isAbsolute(cli.flags.customFormatter)
    ? cli.flags.customFormatter
    : path.join(process.cwd(), cli.flags.customFormatter);
  formatter = dynamicRequire(customFormatter);
}

const optionsBase /*: Object*/ = {
  formatter,
  configOverrides: {}
};

if (cli.flags.quiet) {
  optionsBase.configOverrides.quiet = cli.flags.quiet;
}

if (cli.flags.syntax) {
  optionsBase.syntax = cli.flags.syntax;
}

if (cli.flags.customSyntax) {
  optionsBase.customSyntax = getModulePath(
    process.cwd(),
    cli.flags.customSyntax
  );
}

if (cli.flags.config) {
  // Should check these possibilities:
  //   a. name of a node_module
  //   b. absolute path
  //   c. relative path relative to `process.cwd()`.
  // If none of the above work, we'll try a relative path starting
  // in `process.cwd()`.
  optionsBase.configFile =
    resolveFrom.silent(process.cwd(), cli.flags.config) ||
    path.join(process.cwd(), cli.flags.config);
}

if (cli.flags.configBasedir) {
  optionsBase.configBasedir = path.isAbsolute(cli.flags.configBasedir)
    ? cli.flags.configBasedir
    : path.resolve(process.cwd(), cli.flags.configBasedir);
}

if (cli.flags.stdinFilename) {
  optionsBase.codeFilename = cli.flags.stdinFilename;
}

if (cli.flags.ignorePath) {
  optionsBase.ignorePath = cli.flags.ignorePath;
}

if (cli.flags.ignorePattern) {
  optionsBase.ignorePattern = cli.flags.ignorePattern;
}

if (cli.flags.ignoreDisables) {
  optionsBase.ignoreDisables = cli.flags.ignoreDisables;
}

if (cli.flags.disableDefaultIgnores) {
  optionsBase.disableDefaultIgnores = cli.flags.disableDefaultIgnores;
}

if (cli.flags.cache) {
  optionsBase.cache = true;
}

if (cli.flags.cacheLocation) {
  optionsBase.cacheLocation = cli.flags.cacheLocation;
}

if (cli.flags.fix) {
  optionsBase.fix = cli.flags.fix;
}

const reportNeedlessDisables = cli.flags.reportNeedlessDisables;

if (reportNeedlessDisables) {
  optionsBase.reportNeedlessDisables = reportNeedlessDisables;
}

Promise.resolve()
  .then(() => {
    // Add input/code into options
    if (cli.input.length) {
      return Object.assign({}, optionsBase, {
        files: cli.input
      });
    }
    return getStdin().then(stdin =>
      Object.assign({}, optionsBase, {
        code: stdin
      })
    );
  })
  .then(options => {
    if (!options.files && !options.code) {
      cli.showHelp();
    }

    return standalone(options);
  })
  .then(linted => {
    if (reportNeedlessDisables) {
      const hasReportNeedlessDisable =
        !!linted.needlessDisables &&
        linted.needlessDisables.some(sourceReport => {
          if (!sourceReport.ranges || sourceReport.ranges.length === 0) {
            return false;
          }

          return true;
        });

      if (hasReportNeedlessDisable) {
        process.stdout.write(
          needlessDisablesStringFormatter(linted.needlessDisables)
        );
        process.exitCode = 2;
      }

      return;
    }

    if (!linted.output) {
      return;
    }
    process.stdout.write(linted.output);
    if (linted.errored) {
      process.exitCode = 2;
    }
  })
  .catch(err => {
    console.log(err.stack); // eslint-disable-line no-console
    const exitCode = typeof err.code === "number" ? err.code : 1;
    process.exit(exitCode);
  });
