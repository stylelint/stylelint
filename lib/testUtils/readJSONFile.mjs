import * as fs from 'node:fs';

/**
 * @param {string | URL} filePath
 * @returns {unknown}
 */
export default function readJSONFile(filePath) {
	return JSON.parse(fs.readFileSync(filePath, 'utf8'));
}
