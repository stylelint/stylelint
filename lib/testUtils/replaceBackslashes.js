'use strict';

module.exports = function replaceBackslashes(str) {
	return str.replace(/\\/g, '/');
};
