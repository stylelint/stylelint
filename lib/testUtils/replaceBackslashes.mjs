import { fileURLToPath } from 'node:url';

/**
 * @param {string | URL} path
 * @returns {string}
 */
export default function replaceBackslashes(path) {
	return (typeof path === 'string' ? path : fileURLToPath(path)).replaceAll('\\', '/');
}
