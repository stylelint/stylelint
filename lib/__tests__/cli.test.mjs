/* eslint no-console: off */

import { Readable } from 'node:stream';
import { fileURLToPath } from 'node:url';
import path from 'node:path';

import { jest } from '@jest/globals';
import stripAnsi from 'strip-ansi';
import { stripIndent } from 'common-tags';

import readJSONFile from '../testUtils/readJSONFile.mjs';
import replaceBackslashes from '../testUtils/replaceBackslashes.mjs';

import cli, { buildCLI } from '../cli.mjs';

const pkg = readJSONFile(new URL('../../package.json', import.meta.url));

const fixturesPath = (...elems) =>
	replaceBackslashes(path.join(fileURLToPath(new URL('./fixtures', import.meta.url)), ...elems));

describe('buildCLI', () => {
	it('flags - default', () => {
		expect(buildCLI([]).flags).toEqual({
			allowEmptyInput: false,
			cache: false,
			color: false,
			disableDefaultIgnores: false,
			fix: false,
			formatter: 'string',
			help: false,
			ignoreDisables: false,
			ignorePath: [],
			ignorePattern: [],
			printConfig: false,
			quiet: false,
			quietDeprecationWarnings: false,
			reportDescriptionlessDisables: false,
			reportInvalidScopeDisables: false,
			reportNeedlessDisables: false,
			stdin: false,
			version: false,
		});
	});

	it('flags.allowEmptyInput', () => {
		expect(buildCLI(['--allow-empty-input']).flags.allowEmptyInput).toBe(true);
		expect(buildCLI(['--aei']).flags.allowEmptyInput).toBe(true);
	});

	it('flags.cache', () => {
		expect(buildCLI(['--cache']).flags.cache).toBe(true);
	});

	it('flags.cacheLocation', () => {
		expect(buildCLI(['--cache-location=foo']).flags.cacheLocation).toBe('foo');
	});

	it('flags.globbyOptions', () => {
		expect(buildCLI(['--globby-options={"dot":true}']).flags.globbyOptions).toBe('{"dot":true}');
		expect(buildCLI(['--go={"dot":true}']).flags.globbyOptions).toBe('{"dot":true}');
	});

	it('flags.cacheStrategy', () => {
		expect(buildCLI(['--cache-strategy=content']).flags.cacheStrategy).toBe('content');
		expect(buildCLI(['--cache-strategy=metadata']).flags.cacheStrategy).toBe('metadata');
	});

	it('flags.color', () => {
		expect(buildCLI(['--color']).flags.color).toBe(true);
		expect(buildCLI(['--no-color']).flags.color).toBe(false);
	});

	it('flags.config', () => {
		expect(buildCLI(['--config=/path/to/file']).flags.config).toBe('/path/to/file');
		expect(buildCLI(['-c=/path/to/file']).flags.config).toBe('/path/to/file');
	});

	it('flags.configBasedir', () => {
		expect(buildCLI(['--config-basedir=/path/to/dir']).flags.configBasedir).toBe('/path/to/dir');
	});

	it('flags.customFormatter', () => {
		expect(buildCLI(['--custom-formatter=foo']).flags.customFormatter).toBe('foo');
	});

	it('flags.customSyntax', () => {
		expect(buildCLI(['--custom-syntax=foo']).flags.customSyntax).toBe('foo');
	});

	it('flags.disableDefaultIgnores', () => {
		expect(buildCLI(['--disable-default-ignores']).flags.disableDefaultIgnores).toBe(true);
		expect(buildCLI(['--di']).flags.disableDefaultIgnores).toBe(true);
	});

	it('flags.fix', () => {
		expect(buildCLI(['--fix']).flags.fix).toBe(true);
	});

	it('flags.formatter', () => {
		expect(buildCLI(['--formatter=json']).flags.formatter).toBe('json');
		expect(buildCLI(['-f', 'json']).flags.formatter).toBe('json');
	});

	it('flags.help', () => {
		expect(buildCLI(['--help']).flags.help).toBe(true);
		expect(buildCLI(['-h']).flags.help).toBe(true);
	});

	it('flags.ignoreDisables', () => {
		expect(buildCLI(['--ignore-disables']).flags.ignoreDisables).toBe(true);
		expect(buildCLI(['--id']).flags.ignoreDisables).toBe(true);
	});

	it('flags.ignorePath', () => {
		expect(buildCLI(['--ignore-path=/path/to/file']).flags.ignorePath).toEqual(['/path/to/file']);
		expect(buildCLI(['-i', '/path/to/file']).flags.ignorePath).toEqual(['/path/to/file']);
		expect(
			buildCLI(['--ignore-path=/path/to/file1', '--ignore-path=/path/to/file2']).flags.ignorePath,
		).toEqual(['/path/to/file1', '/path/to/file2']);
		expect(buildCLI(['-i', '/path/to/file1', '-i', '/path/to/file2']).flags.ignorePath).toEqual([
			'/path/to/file1',
			'/path/to/file2',
		]);
	});

	it('flags.ignorePattern', () => {
		expect(buildCLI(['--ignore-pattern=vendor/**']).flags.ignorePattern).toEqual(['vendor/**']);
		expect(buildCLI(['--ip', 'vendor/**']).flags.ignorePattern).toEqual(['vendor/**']);
		expect(buildCLI(['--ip', 'vendor/**', '--ip', 'test/*.css']).flags.ignorePattern).toEqual([
			'vendor/**',
			'test/*.css',
		]);
	});

	it('flags.maxWarnings', () => {
		expect(buildCLI(['--max-warnings=7']).flags.maxWarnings).toBe(7);
		expect(buildCLI(['--mw', '-1']).flags.maxWarnings).toBe(-1);
	});

	it('flags.outputFile', () => {
		expect(buildCLI(['--output-file=/path/to/file']).flags.outputFile).toBe('/path/to/file');
		expect(buildCLI(['-o', '/path/to/file']).flags.outputFile).toBe('/path/to/file');
	});

	it('flags.printConfig', () => {
		expect(buildCLI(['--print-config']).flags.printConfig).toBe(true);
	});

	it('flags.quiet', () => {
		expect(buildCLI(['--quiet']).flags.quiet).toBe(true);
		expect(buildCLI(['-q']).flags.quiet).toBe(true);
	});

	it('flags.quietDeprecationWarnings', () => {
		expect(buildCLI(['--quiet-deprecation-warnings']).flags.quietDeprecationWarnings).toBe(true);
	});

	it('flags.reportInvalidScopeDisables', () => {
		expect(buildCLI(['--report-invalid-scope-disables']).flags.reportInvalidScopeDisables).toBe(
			true,
		);
		expect(buildCLI(['--risd']).flags.reportInvalidScopeDisables).toBe(true);
	});

	it('flags.reportNeedlessDisables', () => {
		expect(buildCLI(['--report-needless-disables']).flags.reportNeedlessDisables).toBe(true);
		expect(buildCLI(['--rd']).flags.reportNeedlessDisables).toBe(true);
	});

	it('flags.stdin', () => {
		expect(buildCLI(['--stdin']).flags.stdin).toBe(true);
	});

	it('flags.stdinFilename', () => {
		expect(buildCLI(['--stdin-filename=foo.css']).flags.stdinFilename).toBe('foo.css');
	});

	it('flags.version', () => {
		expect(buildCLI(['--version']).flags.version).toBe(true);
		expect(buildCLI(['-v']).flags.version).toBe(true);
	});
});

describe('CLI', () => {
	beforeAll(() => {
		jest.spyOn(process, 'exit').mockImplementation(() => {});
		jest.spyOn(process.stdout, 'write').mockImplementation(() => {});
		jest.spyOn(process.stderr, 'write').mockImplementation(() => {});
		jest.spyOn(process, 'stdin', 'get').mockImplementation(() => Readable.from([Buffer.from('')]));
		jest.spyOn(console, 'log').mockImplementation(() => {});
		jest.spyOn(console, 'error').mockImplementation(() => {});
	});

	afterEach(() => {
		process.exitCode = undefined;
	});

	afterAll(() => {
		jest.restoreAllMocks();
	});

	it('basic', async () => {
		await cli([]);

		expect(process.exit).toHaveBeenCalledWith(0);

		expect(console.log).toHaveBeenCalledTimes(1);
		expect(console.log).toHaveBeenCalledWith(
			expect.stringContaining('Usage: stylelint [input] [options]'),
		);
	});

	it('--help', async () => {
		await cli(['--help']);

		expect(process.exit).toHaveBeenCalledWith(0);

		expect(console.log).toHaveBeenCalledTimes(1);
		expect(console.log.mock.calls[0][0]).toMatchSnapshot();
	});

	it('--version', async () => {
		await cli(['--version']);

		expect(process.exitCode).toBeUndefined();

		expect(console.log).toHaveBeenCalledTimes(1);
		expect(console.log).toHaveBeenCalledWith(expect.stringContaining(pkg.version));
	});

	it('--print-config', async () => {
		await cli([
			'--print-config',
			'--config',
			fixturesPath('config-color-no-invalid-hex.json'),
			fixturesPath('invalid-hex.css'),
		]);

		expect(process.exitCode).toBeUndefined();

		expect(process.stdout.write).toHaveBeenCalledTimes(1);
		expect(process.stdout.write).toHaveBeenLastCalledWith(
			JSON.stringify(
				{
					rules: {
						'color-no-invalid-hex': [true],
					},
				},
				null,
				'  ',
			),
		);
	});

	it('--report-needless-disables', async () => {
		await cli([
			'--report-needless-disables',
			'--config',
			fixturesPath('config-block-no-empty.json'),
			fixturesPath('empty-block-with-disables.css'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(
			expect.stringMatching(/Needless disable for "color-named".*Unexpected empty block/s),
		);
	});

	it('reports disallowed disables', async () => {
		await cli([
			'--config',
			fixturesPath('config-block-no-empty-report-disables.json'),
			fixturesPath('empty-block-with-relevant-disable.css'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(
			expect.stringContaining('Rule "block-no-empty" may not be disabled'),
		);
	});

	it('reports descriptionless disables', async () => {
		await cli([
			'--report-descriptionless-disables',
			'--config',
			fixturesPath('config-block-no-empty.json'),
			fixturesPath('empty-block-with-relevant-disable.css'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(
			expect.stringContaining('Disable for "block-no-empty" is missing a description'),
		);
	});

	it('--stdin', async () => {
		await cli(['--stdin', '--config', fixturesPath('config-no-empty-source.json')]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenNthCalledWith(
			1,
			expect.stringContaining('Unexpected empty source'),
		);
	});

	it('exits with non zero on unfound module in config', async () => {
		await cli([
			'--config',
			fixturesPath('config-require-unknown.js'),
			fixturesPath('empty-block-with-disables.css'),
		]);

		expect(process.exitCode).toBe(1);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(
			expect.stringContaining('Cannot find module'),
		);
	});

	it('outputs a valid JSON to stderr even if max warnings exceeded', async () => {
		await cli([
			'--formatter=json',
			'--max-warnings=0',
			'--config',
			fixturesPath('default-severity-warning.json'),
			fixturesPath('empty-block.css'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(2);
		const output = JSON.parse(process.stderr.write.mock.calls[0][0]);

		expect(output).toBeInstanceOf(Array);
		expect(output[0]).toHaveProperty('source', expect.stringContaining('empty-block.css'));

		expect(stripAnsi(process.stderr.write.mock.calls[1][0])).toContain(
			'Max warnings exceeded: 1 found. 0 allowed',
		);
	});

	it('--quiet', async () => {
		await cli([
			'--quiet',
			'--config',
			fixturesPath('default-severity-warning.json'),
			fixturesPath('empty-block.css'),
		]);

		expect(process.exitCode).toBeUndefined();

		expect(process.stdout.write).toHaveBeenCalledTimes(0);
		expect(process.stderr.write).toHaveBeenCalledTimes(0);
	});

	it('--quiet-deprecation-warnings', async () => {
		await cli([
			'--quiet-deprecation-warnings',
			'--config',
			fixturesPath('quiet-deprecation-warnings/config.json'),
			fixturesPath('quiet-deprecation-warnings/style.css'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).toHaveBeenCalledTimes(0);
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write.mock.calls[0][0]).toEqual(
			expect.not.stringContaining('The "color-hex-case" rule is deprecated.'),
		);
	});

	it('--custom-formatter with filesystem path', async () => {
		await cli([
			`--custom-formatter=${fixturesPath('formatters/customFormatter')}`,
			'--config',
			fixturesPath('default-severity-warning.json'),
			fixturesPath('empty-block.css'),
		]);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith('Custom formatter reports 1 warning(s).');
	});

	it('--custom-formatter with node module', async () => {
		await cli([
			`--custom-formatter=stylelint-test-custom-formatter`,
			'--config',
			fixturesPath('default-severity-warning.json'),
			fixturesPath('empty-block.css'),
		]);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith('Custom formatter reports 1 warning(s).');
	});

	it('output a message when wrong --globby-options provided', async () => {
		await cli(['--globby-options=wrong']);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stdout.write).toHaveBeenCalledTimes(0);
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(stripAnsi(process.stderr.write.mock.calls[0][0])).toContain(
			'Invalid option "--globby-options". The value "wrong" is not valid JSON object.',
		);
	});

	it('--globby-options', async () => {
		await cli([
			'--globby-options={"dot":true}',
			'--config',
			fixturesPath('config-block-no-empty.json'),
			fixturesPath('globby-options'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(expect.stringMatching(/block-no-empty/));
	});

	it('--custom-syntax', async () => {
		await cli([
			'--custom-syntax=postcss-scss',
			'--config',
			fixturesPath('config-color-no-invalid-hex.json'),
			fixturesPath('invalid-hex.scss'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(
			expect.stringMatching(/color-no-invalid-hex/),
		);
	});

	it('--custom-syntax and --config-basedir', async () => {
		await cli([
			'--custom-syntax=./custom-syntax',
			'--config-basedir',
			fixturesPath(),
			'--config',
			fixturesPath('config-color-no-invalid-hex.json'),
			fixturesPath('invalid-hex.scss'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(
			expect.stringMatching(/color-no-invalid-hex/),
		);
	});

	it("doesn't write to stdout or stderr when linting a file without errors", async () => {
		await cli([
			'--config',
			fixturesPath('config-block-empty-ok.json'),
			fixturesPath('empty-block.css'),
		]);

		expect(process.exitCode).toBeUndefined();

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).not.toHaveBeenCalled();
	});

	it('writes to stderr when linting a file with errors', async () => {
		await cli([
			'--config',
			fixturesPath('config-block-no-empty.json'),
			fixturesPath('empty-block.css'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).not.toHaveBeenCalled();
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(expect.stringMatching(/block-no-empty/));
	});

	it('writes to stdout when linting code from stdin that can be fully fixed', async () => {
		jest
			.spyOn(process, 'stdin', 'get')
			.mockImplementationOnce(() => Readable.from([Buffer.from('a { color: #ffffff; }')]));
		await cli([
			'--stdin',
			'--fix',
			'--config',
			fixturesPath('config-block-no-empty-and-color-hex-length-short.json'),
		]);

		expect(process.exitCode).toBeUndefined();

		expect(process.stdout.write).toHaveBeenCalledTimes(1);
		expect(process.stdout.write).toHaveBeenCalledWith('a { color: #fff; }');
		expect(process.stderr.write).not.toHaveBeenCalled();
	});

	it('writes to stdout and stderr when linting code from stdin that can only be partially fixed', async () => {
		jest.spyOn(process, 'stdin', 'get').mockImplementationOnce(() =>
			Readable.from([
				Buffer.from(stripIndent`
				a { color: #ffffff; }
				div {}
			`),
			]),
		);
		await cli([
			'--stdin',
			'--fix',
			'--config',
			fixturesPath('config-block-no-empty-and-color-hex-length-short.json'),
		]);

		expect(process.exitCode).toBe(2);

		expect(process.stdout.write).toHaveBeenCalledTimes(1);
		expect(process.stdout.write).toHaveBeenCalledWith(stripIndent`
			a { color: #fff; }
			div {}
		`);
		expect(process.stderr.write).toHaveBeenCalledTimes(1);
		expect(process.stderr.write).toHaveBeenCalledWith(expect.stringMatching(/block-no-empty/));
	});
});
