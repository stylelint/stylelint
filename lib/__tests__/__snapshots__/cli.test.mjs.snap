// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI --help 1`] = `
"
  A mighty CSS linter that helps you avoid errors and enforce conventions.

  Usage: stylelint [input] [options]

  Input: Files(s), glob(s), or nothing to use stdin.

    If an input argument is wrapped in quotation marks, it will be passed to
    globby for cross-platform glob support. "node_modules" are always ignored.
    You can also pass no input and use stdin, instead.

  Options:

    --config, -c <path_or_module>

      A path to a specific configuration file (JSON, YAML, CommonJS, or ES module),
      or a module name in "node_modules" that points to one. If no argument is
      provided, Stylelint will search for configuration files in the following
      places, in this order:

        - a "stylelint" property in "package.json"
        - a ".stylelintrc" file
        - a ".stylelintrc.{cjs,mjs,js,json,yaml,yml}" file
        - a "stylelint.config.{cjs,mjs,js}" file

      The search will begin in the working directory and move up the directory
      tree until a configuration file is found.

    --config-basedir <path>

      An absolute path to the directory that relative paths defining "extends",
      "plugins", and "customSyntax" are *relative to*. Only necessary if these
      values are relative paths.

    --print-config

      Print the configuration for the given input file path. Globs are unsupported.

    --ignore-path, -i <path>

      A path to a file containing patterns that describe files to ignore. The
      path can be absolute or relative to "process.cwd()". You can repeat the
      option to provide multiple paths. By default, Stylelint looks for
      ".stylelintignore" in "process.cwd()". Multiple can be set.

    --ignore-pattern, --ip <pattern>

      A pattern of files to ignore (in addition to those in ".stylelintignore").
      Multiple can be set.

    --fix

      Automatically fix problems of certain rules.

    --custom-syntax <name_or_path>

      A module name or path to a JS file exporting a PostCSS-compatible syntax.

    --stdin

      Accept stdin input even if it is empty.

    --stdin-filename <name>

      A filename to assign stdin input.

    --ignore-disables, --id

      Ignore "stylelint-disable" comments.

    --disable-default-ignores, --di

      Allow linting of "node_modules".

    --[no-]cache

      Store the info about processed files in order to only operate on the
      changed ones the next time you run Stylelint. By default, the cache is
      stored in "./.stylelintcache". To adjust this, use "--cache-location".
      Cache is disabled by default.

    --cache-location <path>

      A path to a file or directory to be used for the cache location. If a
      directory is specified, a cache file will be created inside the specified
      folder, with a name derived from a hash of the current working directory.

      If the directory for the cache does not exist, make sure you add a trailing "/"
      on *nix systems or "\\" on Windows. Otherwise the path will be assumed to
      be a file.

    --cache-strategy <strategy>

      A strategy for the cache to use for detecting changed files. Either one of:

        metadata     by metadata of a file (default)
        content      by content of a file

      The "content" strategy can be useful in cases where the modification time
      of your files changes even if their contents have not. For example, this can
      happen during git operations like "git clone" because git does not track file
      modification time.

    --formatter, -f <formatter>

      An output formatter. The variants are as follows:

        string       human-readable strings (default)
        compact      similar to ESLint's compact formatter
        github       workflow commands for GitHub Actions (DEPRECATED)
        json         JSON format
        tap          TAP format
        unix         C compiler-like format
        verbose      extend "string" to include a file list and a tally

    --custom-formatter <path_or_module>

      A path to a JS file or module name exporting a custom formatting function.

    --quiet, -q

      Only register problems for rules with an "error"-level severity (ignore
      "warning"-level).

    --quiet-deprecation-warnings

      Ignore deprecations warnings.

    --[no-]color

      Force enabling/disabling of color.

    --report-needless-disables, --rd

      Also report errors for "stylelint-disable" comments that are not blocking
      a lint warning. The process will exit with code 2 if needless disables are found.

    --report-invalid-scope-disables, --risd

      Report "stylelint-disable" comments that used for rules that don't exist
      within the configuration object. The process will exit with code 2 if invalid
      scope disables are found.

    --report-descriptionless-disables, --rdd

      Report "stylelint-disable" comments without a description. The process will
      exit with code 2 if descriptionless disables are found.

    --max-warnings, --mw <number>

      The number of warnings above which the process will exit with code 2.
      Useful when setting "defaultSeverity" to "warning" and expecting the process
      to fail on warnings (e.g. CI build).

    --output-file, -o <path>

      A file path to write a report.

    --allow-empty-input, --aei

      When a glob pattern matches no files, the process will exit without throwing an error.

    --globby-options, --go <json>

      Options in JSON format passed to globby.

    --version, -v

      Show the version.

    --help, -h

      Show the help.
"
`;
